wren-ai-service/src/web/v1/services/question_recommendation.py

```diff
@@ -0,0 +1,135 @@
+        error: Optional[Error] = None
+
+    def __init__(
+        self,
+        pipelines: Dict[str, BasicPipeline],
+        maxsize: int = 1_000_000,
+        ttl: int = 120,
+    ):
+        self._pipelines = pipelines
+        self._cache: Dict[str, QuestionRecommendation.Resource] = TTLCache(
+            maxsize=maxsize, ttl=ttl
+        )
+
+    def _handle_exception(
+        self,
+        input: Input,
+        error_message: str,
+        code: str = "OTHERS",
+    ):
+        self._cache[input.id] = self.Resource(
+            id=input.id,
+            status="failed",
+            error=self.Resource.Error(code=code, message=error_message),
+        )
+        logger.error(error_message)
+
+    async def _validate_question(
+        self,
+        candidate: dict,
+        project_id: Optional[str] = None,
+    ) -> bool:
+        print(f"Question: {candidate}")
+        retrieval_result = await self._pipelines["retrieval"].run(
+            query=candidate["question"],
+            id=project_id,
+        )
+        documents = retrieval_result.get("construct_retrieval_results", [])
+        generated_sql = await self._pipelines["sql_generation"].run(
+            query=candidate["question"],
+            contexts=documents,
+            exclude=[],
+            configurations=AskConfigurations(),
+        )
+        valid_sql = generated_sql["post_process"]["valid_generation_results"]
+        logger.debug(f"Valid SQL: {valid_sql}")
+
+        return True if valid_sql else False
+
+    @observe(name="Generate Question Recommendation")
+    @trace_metadata
+    async def recommend(self, request: Input, **kwargs) -> Resource:
+        logger.info("Generate Question Recommendation pipeline is running...")
+
+        try:
+            input = {
+                "mdl": orjson.loads(request.mdl),
+                "previous_questions": request.previous_questions,
+                "language": request.language,
+            }
+
+            resp = await self._pipelines["question_recommendation"].run(**input)
+            questions = resp.get("normalized", {}).get("questions", [])
+
+            validated_questions = [
+                question
+                for question in questions
+                if await self._validate_question(question, request.project_id)
+            ]
+
+            self._cache[request.id] = self.Resource(
+                id=request.id,
+                status="finished",
+                response={"questions": validated_questions},
+            )
+        except orjson.JSONDecodeError as e:
+            self._handle_exception(
+                request,
+                f"Failed to parse MDL: {str(e)}",
+                code="MDL_PARSE_ERROR",
+            )
+        except Exception as e:
+            self._handle_exception(
+                request,
+                f"An error occurred during question recommendation generation: {str(e)}",
+            )
+
+        return self._cache[request.id]
```

Comment:
I wonder if some metadata like whether error type and error message will be recorded in Langfuse?